File: crates/oracle/h2-6a-map-option-projection.mjs
// Independent option-projection witnesses; never rewrites qualification inputs.
import assert from "node:assert/strict";
import fs from "node:fs";
import path from "node:path";
import { createHash } from "node:crypto";
import { createRequire } from "node:module";
import { fileURLToPath } from "node:url";

const workspace = fileURLToPath(new URL("../../", import.meta.url));
const require = createRequire(import.meta.url);
const ts = require(path.join(workspace, "vendor/typescript-6.0.3/lib/typescript.js"));
assert.equal(ts.version, "6.0.3");
const hash = (bytes) => createHash("sha256").update(bytes).digest("hex");
const bytes = (text) => ({ utf8_base64: Buffer.from(text).toString("base64"), utf8_sha256: hash(text), utf8_bytes: Buffer.byteLength(text) });
const mapNames = ["sourceMap", "inlineSourceMap", "inlineSources", "sourceRoot", "mapRoot"];
const fixturePath = path.join(workspace, "crates/compiler/tests/fixtures/h2-6a-map-option-projection.json");
const inventoryPath = path.join(workspace, "docs/design/greenfield/slices/h2-6a-map-option-inventory.json");
const qualificationPath = "ratchets/h2-6a-qualification.v1.json";
const qualificationBytes = fs.readFileSync(path.join(workspace, qualificationPath));
const qualification = JSON.parse(qualificationBytes);

function inventory() {
  const cases = qualification.cases.map((row) => ({
    case_id: row.case_id,
    execution_route: row.execution_route,
    disposition: row.disposition,
    directive_map_options: Object.fromEntries(row.input.settings.filter(({ name }) => mapNames.some((key) => key.toLowerCase() === name.toLowerCase())).map(({ name, value }) => [name.toLowerCase(), value])),
    virtual_config: row.input.virtual_config,
    frozen_typescript_runs: row.typescript_run_fingerprints,
  }));
  assert.equal(cases.length, 177);
  assert.equal(cases.filter((row) => row.virtual_config !== null).length, 0);
  const extra = cases.filter((row) => Object.keys(row.directive_map_options).some((key) => key !== "sourcemap"));
  assert.equal(extra.length, 3);
  return { schema: 1, purpose: "Input inventory; no native parity claim", qualification: qualificationPath, qualification_sha256: hash(qualificationBytes), count: cases.length, additional_map_option_rows: extra.map((row) => row.case_id), cases };
}

const variants = [
  ["external", { sourceMap: true }],
  ["external-sources", { sourceMap: true, inlineSources: true }],
  ["relative-map-root", { sourceMap: true, inlineSources: true, mapRoot: "local" }],
  ["relative-source-root", { sourceMap: true, inlineSources: true, sourceRoot: "local" }],
  ["empty-roots", { sourceMap: true, inlineSources: true, mapRoot: "", sourceRoot: "" }],
  ["false-inline-sources", { sourceMap: true, inlineSources: false }],
  ["false-inline-map", { sourceMap: true, inlineSourceMap: false }],
  ["all-false", { sourceMap: false, inlineSourceMap: false, inlineSources: false, mapRoot: "", sourceRoot: "" }],
  ["inline", { sourceMap: false, inlineSourceMap: true, inlineSources: true }],
  ["inline-empty-roots", { inlineSourceMap: true, inlineSources: true, mapRoot: "", sourceRoot: "" }],
  ["conflicting-maps", { sourceMap: true, inlineSourceMap: true }],
  ["sources-without-map", { inlineSources: true }],
  ["map-root-without-map", { mapRoot: "local" }],
  ["source-root-without-map", { sourceRoot: "local" }],
  ["map-root-with-inline", { inlineSourceMap: true, mapRoot: "local" }],
];

function diagnostic(d) {
  return { code: d.code, category: ts.DiagnosticCategory[d.category], file: d.file?.fileName ?? null, start: d.start ?? null, length: d.length ?? null, message: ts.flattenDiagnosticMessageText(d.messageText, "\n") };
}
function observe(input) {
  const files = new Map(input.files.map((file) => [file.path, Buffer.from(file.utf8_base64, "base64").toString("utf8")]));
  if (input.virtual_config) files.set(input.virtual_config.path, Buffer.from(input.virtual_config.utf8_base64, "base64").toString("utf8"));
  const cwd = input.current_directory;
  const normalize = (name) => ts.getNormalizedAbsolutePath(name, cwd);
  const readFile = (name) => files.get(normalize(name)) ?? ts.sys.readFile(name);
  const fileExists = (name) => files.has(normalize(name)) || ts.sys.fileExists(name);
  let options = { noResolve: false };
  if (input.virtual_config) {
    const configPath = input.virtual_config.path;
    const config = ts.parseJsonText(configPath, readFile(configPath));
    const parsed = ts.parseJsonSourceFileConfigFileContent(config, { useCaseSensitiveFileNames: true, readFile, fileExists, readDirectory: () => input.roots }, path.posix.dirname(configPath), undefined, configPath);
    assert.deepEqual(parsed.errors, []);
    options = parsed.options;
  }
  Object.assign(options, { newLine: ts.NewLineKind.CarriageReturnLineFeed, noErrorTruncation: true, skipDefaultLibCheck: true });
  for (const { name, value } of input.settings) {
    const option = ts.optionDeclarations.find((option) => option.name.toLowerCase() === name.toLowerCase());
    assert.ok(option, name);
    const errors = [];
    options[option.name] = option.type === "boolean" ? value.toLowerCase() === "true" : option.type === "string" ? value : ts.parseCustomTypeOption(option, value, errors);
    assert.deepEqual(errors, []);
  }
  const base = ts.createCompilerHost(options, true);
  const host = { ...base, getCurrentDirectory: () => cwd, getCanonicalFileName: (name) => name, useCaseSensitiveFileNames: () => true, readFile, fileExists,
    directoryExists: (dir) => [...files.keys()].some((name) => name.startsWith(`${normalize(dir).replace(/\/$/, "")}/`)) || base.directoryExists(dir),
    getSourceFile(name, version) { const text = files.get(normalize(name)); return text === undefined ? base.getSourceFile(name, version) : ts.createSourceFile(name, text, version, true); },
  };
  const program = ts.createProgram(input.roots, options, host);
  const writes = [], reported = [], status = [];
  let result;
  const original = program.emit;
  program.emit = function (...args) { result = original.apply(this, args); return result; };
  const exit = ts.emitFilesAndReportErrorsAndGetExitStatus(program, (d) => reported.push(diagnostic(d)), (text) => status.push(text), undefined,
    (name, text, bom, onError, sourceFiles, data) => {
      const callback = Buffer.from(text), materialized = bom ? Buffer.concat([Buffer.from([239, 187, 191]), callback]) : callback;
      writes.push({ index: writes.length, path: name, kind: name.endsWith(".map") ? "source-map" : "javascript", callback_utf8_base64: callback.toString("base64"), callback_utf8_sha256: hash(callback), callback_utf8_bytes: callback.length, write_byte_order_mark: bom, materialized_utf8_sha256: hash(materialized), materialized_utf8_bytes: materialized.length, on_error_callback_present: onError !== undefined, source_files: (sourceFiles ?? []).map((source) => source.fileName), data_present: data !== undefined, data_source_map_url_pos: data?.sourceMapUrlPos ?? null, data_diagnostics_count: data?.diagnostics?.length ?? null });
    });
  return { effective_map_options: Object.fromEntries(mapNames.map((name) => [name, options[name] ?? null])), observation: { writes, reported_diagnostics: reported, emit_result: { emit_skipped: result.emitSkipped, diagnostics: result.diagnostics.map(diagnostic), emitted_files: result.emittedFiles ?? null, source_maps: result.sourceMaps?.map((entry) => ({ input_source_file_names: entry.inputSourceFileNames, source_map_json: JSON.stringify(entry.sourceMap) })) ?? null }, status_writes: status, exit_code: exit } };
}

function build() {
  const cases = [];
  for (const route of ["directives", "virtual-config"]) {
    for (const [slug, mapOptions] of variants) {
      const config = route === "virtual-config" ? { compilerOptions: { target: "es2015", ...mapOptions }, files: ["input.ts"] } : null;
      const input = { current_directory: "/.src", roots: ["/.src/input.ts"], files: [{ path: "/.src/input.ts", ...bytes("var a = 10;") }], settings: config ? [] : Object.entries({ target: "es2015", ...mapOptions }).map(([name, value]) => ({ name, value: String(value) })), virtual_config: config ? { path: "/.src/tsconfig.json", ...bytes(JSON.stringify(config, null, 2)) } : null };
      const first = observe(input), second = observe(input);
      assert.deepEqual(first, second);
      cases.push({ case_id: `${route}/${slug}`, route, input, repetitions: 2, ...first });
    }
  }
  // A higher directive layer must retain explicit false and empty strings.
  const inherited = structuredClone(cases.find((row) => row.case_id === "virtual-config/relative-map-root").input);
  inherited.settings = [{ name: "inlineSources", value: "false" }, { name: "mapRoot", value: "" }, { name: "sourceRoot", value: "" }];
  const first = observe(inherited);
  assert.deepEqual(first, observe(inherited));
  cases.push({ case_id: "virtual-config/directive-overrides", route: "virtual-config", input: inherited, repetitions: 2, ...first });
  return { schema: 1, typescript: ts.version, generator_sha256: hash(fs.readFileSync(fileURLToPath(import.meta.url))), typescript_sha256: hash(fs.readFileSync(path.join(workspace, "vendor/typescript-6.0.3/lib/typescript.js"))), cases };
}
const mode = process.argv[2];
assert.ok(["--write", "--check", "--inventory"].includes(mode), "usage: --write | --check | --inventory");
const render = (value) => `${JSON.stringify(value, null, 2)}\n`;
if (mode === "--inventory") {
  fs.writeFileSync(inventoryPath, render(inventory()));
  console.log("inventoried 177 H2.6a inputs; 3 additional map option rows; 0 virtual configs");
} else {
  const result = render(build());
  if (mode === "--write") fs.writeFileSync(fixturePath, result);
  else assert.equal(fs.readFileSync(fixturePath, "utf8"), result);
  console.log(`${mode}: 31 option witnesses, two TypeScript observations each`);
}

File: crates/compiler/tests/h2_6a_map_option_projection.rs
//! Compare complete frozen observations through the original H2.6a routes.
//! This target also exposes the unchanged investigation probe without adding
//! another registration to the shared contracts target.
#[path = "integration/source_map_band_probe.rs"]
mod source_map_band_probe;

use std::path::{Path, PathBuf};
use std::sync::Arc;

use base64::Engine as _;
use serde_json::{json, Value};
use sha2::{Digest, Sha256};
use tsc_compiler::{EmitWriteMetadata, H2RuntimeSlice, MemoryOutputSink, ProgramSession};
use tsc_diagnostics::{Diagnostic, MessageChain};
use tsc_harness::upstream_suites::execution::{
    load_compiler_emit_with_option_floor, load_qualified_compiler_emit_with_option_floor,
    load_recorded_execution_plans, CompilerExecutionPlan, CompilerExplicitRootReason,
    CompilerRootSelection, CompilerUnitId, CompilerUnitInput, EmitOptionFloor,
    UpstreamExecutionCorpus, UpstreamExecutionInput,
};
use tsc_harness::upstream_suites::OrderedSetting;
use tsc_host::MemoryCompilerHost;
use tsc_program::{
    parse_config_root_plan, CompilerConfigHost, ConfigRootPlanRequest, PreparedProgram,
    ProgramLoadLimits,
};

const ROWS: &[&str] = &[
    "typescript-6.0.3/compiler/optionsSourcemapInlineSources.ts#default",
    "typescript-6.0.3/compiler/optionsSourcemapInlineSourcesMapRoot.ts#default",
    "typescript-6.0.3/compiler/optionsSourcemapInlineSourcesSourceRoot.ts#default",
];

fn workspace() -> PathBuf {
    Path::new(env!("CARGO_MANIFEST_DIR")).join("../..")
}
fn limits() -> ProgramLoadLimits {
    ProgramLoadLimits::new(256, 2_048, 64, 16 * 1_024 * 1_024, 128 * 1_024 * 1_024)
}
fn artifact(path: &str) -> Value {
    serde_json::from_slice(&std::fs::read(workspace().join(path)).unwrap()).unwrap()
}
fn decode(file: &Value) -> Vec<u8> {
    let bytes = base64::engine::general_purpose::STANDARD
        .decode(file["utf8_base64"].as_str().unwrap())
        .unwrap();
    assert_eq!(sha256(&bytes), file["utf8_sha256"]);
    assert_eq!(bytes.len() as u64, file["utf8_bytes"].as_u64().unwrap());
    bytes
}
fn sha256(bytes: impl AsRef<[u8]>) -> String {
    format!("{:x}", Sha256::digest(bytes.as_ref()))
}
fn settings(input: &Value) -> Vec<(String, String)> {
    input["settings"]
        .as_array()
        .unwrap()
        .iter()
        .map(|setting| {
            (
                setting["name"].as_str().unwrap().to_owned(),
                setting["value"].as_str().unwrap().to_owned(),
            )
        })
        .collect()
}
fn input_files(input: &Value) -> Vec<(PathBuf, Vec<u8>)> {
    let mut files: Vec<_> = input["files"]
        .as_array()
        .unwrap()
        .iter()
        .map(|file| (PathBuf::from(file["path"].as_str().unwrap()), decode(file)))
        .collect();
    if !input["virtual_config"].is_null() {
        let config = &input["virtual_config"];
        files.push((
            PathBuf::from(config["path"].as_str().unwrap()),
            decode(config),
        ));
    }
    files
}
fn qualified(input: &Value, floor: EmitOptionFloor) -> PreparedProgram {
    load_qualified_compiler_emit_with_option_floor(
        &workspace(),
        input["current_directory"].as_str().unwrap(),
        &input_files(input),
        &input["roots"]
            .as_array()
            .unwrap()
            .iter()
            .map(|root| PathBuf::from(root.as_str().unwrap()))
            .collect::<Vec<_>>(),
        &settings(input),
        limits(),
        floor,
    )
    .unwrap()
}
fn recorded<'a>(case: &Value, corpus: &'a UpstreamExecutionCorpus) -> &'a CompilerExecutionPlan {
    let row = corpus
        .plans
        .iter()
        .find(|row| row.provenance.case_id.as_ref() == case["case_id"].as_str().unwrap())
        .unwrap();
    assert_eq!(
        u64::from(row.provenance.case_index),
        case["expansion_case"].as_u64().unwrap()
    );
    let UpstreamExecutionInput::Compiler(plan) = &row.input else {
        panic!("compiler plan")
    };
    plan
}
fn prepare(
    case: &Value,
    corpus: &UpstreamExecutionCorpus,
    floor: EmitOptionFloor,
) -> PreparedProgram {
    match case["execution_route"].as_str().unwrap() {
        "recorded-compiler-plan" => load_compiler_emit_with_option_floor(
            &workspace(),
            recorded(case, corpus),
            limits(),
            floor,
        )
        .unwrap(),
        "qualified-vfs" => qualified(&case["input"], floor),
        route => panic!("unexpected H2.6a route {route}"),
    }
}

// Synthetic witnesses use the actual compiler-plan adapter as well as the
// qualified adapter. Only test input construction is synthetic; options are
// parsed by the production config planner and projected by the harness.
fn witness_plan(input: &Value, template: &CompilerExecutionPlan) -> CompilerExecutionPlan {
    let mut plan = template.clone();
    let mut fixture = (*plan.fixture).clone();
    let files = input_files(input);
    fixture.units = files
        .iter()
        .enumerate()
        .map(|(index, (path, bytes))| CompilerUnitInput {
            id: CompilerUnitId(index as u32),
            name: Arc::from(path.to_str().unwrap()),
            content: Some(Arc::from(std::str::from_utf8(bytes).unwrap())),
            file_options: Arc::from([]),
            original_fixture_path: Arc::from("map-option-witness.ts"),
            references: Arc::from([]),
            document_symlinks: Arc::from([]),
        })
        .collect::<Vec<_>>()
        .into();
    fixture.config_unit = None;
    fixture.config_root_plan = None;
    if !input["virtual_config"].is_null() {
        let config = &input["virtual_config"];
        let mut builder = MemoryCompilerHost::builder(input["current_directory"].as_str().unwrap());
        for (path, bytes) in &files {
            builder = builder.file(path, bytes.clone());
        }
        let host = builder.build().unwrap();
        let parsed = parse_config_root_plan(
            &CompilerConfigHost::new(&host),
            ConfigRootPlanRequest {
                file_name: config["path"].as_str().unwrap().to_owned(),
                text: String::from_utf8(decode(config)).unwrap(),
                base_path: input["current_directory"].as_str().unwrap().to_owned(),
            },
        )
        .unwrap();
        fixture.config_unit = Some(CompilerUnitId(1));
        fixture.config_root_plan = Some(Arc::new(parsed));
    }
    plan.fixture = Arc::new(fixture);
    plan.current_directory = Arc::from(input["current_directory"].as_str().unwrap());
    plan.effective_settings = settings(input)
        .into_iter()
        .map(|(name, value)| OrderedSetting { name, value })
        .collect::<Vec<_>>()
        .into();
    plan.root_selection = CompilerRootSelection::Explicit {
        reason: CompilerExplicitRootReason::AllUnits,
        root_units: Arc::from([CompilerUnitId(0)]),
        other_units: Arc::from([]),
        vfs_write_order: (0..files.len())
            .map(|id| CompilerUnitId(id as u32))
            .collect::<Vec<_>>()
            .into(),
        program_root_units: Arc::from([CompilerUnitId(0)]),
    };
    plan
}

fn message(chain: &MessageChain, indent: usize, text: &mut String) {
    if indent != 0 {
        text.push('\n');
        text.push_str(&"  ".repeat(indent));
    }
    text.push_str(&chain.text);
    for next in &chain.next {
        message(next, indent + 1, text);
    }
}
fn diagnostics(list: &[Diagnostic]) -> Value {
    json!(list.iter().map(|d| {
        let mut text = String::new(); message(&d.message, 0, &mut text);
        json!({ "code": d.code(), "category": format!("{:?}", d.category()), "file": d.file_name,
            "start": d.start, "length": d.length, "message": text })
    }).collect::<Vec<_>>())
}
fn observe(prepared: PreparedProgram) -> Value {
    let options = prepared.compiler_options();
    let effective = json!({ "sourceMap": options.source_map, "inlineSourceMap": options.inline_source_map,
        "inlineSources": options.inline_sources, "sourceRoot": options.source_root, "mapRoot": options.map_root });
    let session = ProgramSession::new(prepared);
    let bundle = session.prepare_harness_lib_bundle().unwrap();
    let mut sink = MemoryOutputSink::new();
    let (outcome, reported) = session
        .emit_with_reported_diagnostics_for_harness_with_lib_bundle(&mut sink, bundle.as_ref())
        .unwrap();
    let writes = sink.writes().iter().enumerate().map(|(index, write)| {
        let (data_present, pos, count) = match write.metadata() {
            None => (false, None, None),
            Some(EmitWriteMetadata::Text(data)) => (true, data.source_map_url_position().map(|pos| pos.value()), Some(data.diagnostics().len())),
            _ => panic!("unexpected build info"),
        };
        json!({ "index": index, "path": write.path().to_string_lossy(),
            "kind": if write.path().extension().is_some_and(|ext| ext == "map") { "source-map" } else { "javascript" },
            "callback_utf8_base64": base64::engine::general_purpose::STANDARD.encode(write.callback_bytes()),
            "callback_utf8_sha256": sha256(write.callback_bytes()), "callback_utf8_bytes": write.callback_bytes().len(),
            "write_byte_order_mark": write.write_byte_order_mark(), "materialized_utf8_sha256": sha256(write.materialized_bytes()),
            "materialized_utf8_bytes": write.materialized_bytes().len(), "on_error_callback_present": true,
            "source_files": write.source_files().map(|files| files.iter().map(|path| path.to_string_lossy()).collect::<Vec<_>>()),
            "data_present": data_present, "data_source_map_url_pos": pos, "data_diagnostics_count": count })
    }).collect::<Vec<_>>();
    json!({ "effective_map_options": effective,
        "map_runtime_activity": {
            "H2.6a": outcome.h2_activity().runtime_slice(H2RuntimeSlice::H2_6a),
            "H2.6b": outcome.h2_activity().runtime_slice(H2RuntimeSlice::H2_6b),
            "H2.6c": outcome.h2_activity().runtime_slice(H2RuntimeSlice::H2_6c),
        }, "observation": {
        "writes": writes, "reported_diagnostics": diagnostics(&reported),
        "emit_result": { "emit_skipped": outcome.emit_skipped(), "diagnostics": diagnostics(outcome.diagnostics()),
            "emitted_files": outcome.emitted_files().map(|paths| paths.iter().map(|path| path.to_string_lossy()).collect::<Vec<_>>()),
            "source_maps": outcome.source_maps().map(|maps| maps.iter().map(|map| json!({
                "input_source_file_names": map.input_source_files().iter().map(|path| path.to_string_lossy()).collect::<Vec<_>>(),
                "source_map_json": map.canonical_json() })).collect::<Vec<_>>()) },
        "status_writes": [], "exit_code": if reported.is_empty() { 0 } else if outcome.emit_skipped() { 1 } else { 2 }
    } })
}
fn expected(case: &Value) -> Value {
    let mut expected = case["typescript_observation"].clone();
    expected
        .as_object_mut()
        .unwrap()
        .remove("run_fingerprint_sha256");
    expected
}
fn capture(id: &str, floor: EmitOptionFloor, attempt: usize, actual: &Value, expected: &Value) {
    if let Some(out) = std::env::var_os("TSRS_MAP_OPTION_CAPTURE") {
        let out = PathBuf::from(out);
        std::fs::create_dir_all(&out).unwrap();
        std::fs::write(out.join(format!("{}-{floor:?}-{attempt}.json", id.replace(['/', '#', '%'], "_"))),
            serde_json::to_vec_pretty(&json!({ "case_id": id, "floor": format!("{floor:?}"), "attempt": attempt, "actual": actual, "expected": expected })).unwrap()).unwrap();
    }
}

#[test]
fn original_floor_divergence_and_existing_map_family_parity() {
    let qualification = artifact("ratchets/h2-6a-qualification.v1.json");
    let corpus = load_recorded_execution_plans(&workspace()).unwrap();
    for (ordinal, id) in ROWS.iter().enumerate() {
        let case = qualification["cases"]
            .as_array()
            .unwrap()
            .iter()
            .find(|case| case["case_id"] == *id)
            .unwrap();
        let expected = expected(case);
        for floor in [EmitOptionFloor::SourceMap, EmitOptionFloor::MapFamily] {
            let mut runs = Vec::new();
            for attempt in 0..2 {
                let actual = observe(prepare(case, &corpus, floor));
                capture(id, floor, attempt, &actual, &expected);
                runs.push(actual);
            }
            assert_eq!(runs[0], runs[1], "{id}: {floor:?} determinism");
            let actual = &runs[0]["observation"];
            if floor == EmitOptionFloor::MapFamily {
                assert_eq!(
                    actual, &expected,
                    "{id}: existing native map option implementation"
                );
            } else {
                assert_eq!(
                    actual["reported_diagnostics"],
                    expected["reported_diagnostics"]
                );
                assert_eq!(actual["exit_code"], expected["exit_code"]);
                assert_ne!(actual["emit_result"], expected["emit_result"]);
                let differing = actual["writes"]
                    .as_array()
                    .unwrap()
                    .iter()
                    .zip(expected["writes"].as_array().unwrap())
                    .filter(|(a, b)| a != b)
                    .count();
                assert_eq!(differing, if ordinal == 1 { 2 } else { 1 });
                assert_eq!(
                    runs[0]["effective_map_options"]["inlineSources"],
                    Value::Null
                );
            }
            eprintln!("{id}: {floor:?} verified x2");
        }
    }
}

#[test]
fn h2_6a_rows_and_adjacent_controls_match_complete_frozen_tuples() {
    let qualification = artifact("ratchets/h2-6a-qualification.v1.json");
    let corpus = load_recorded_execution_plans(&workspace()).unwrap();
    let controls = [
        "typescript-6.0.3/compiler/sourceMapValidationVariables.ts#default",
        "typescript-6.0.3/conformance/es6/computedProperties/computedPropertyNamesSourceMap1_ES6.ts#default",
    ];
    for id in ROWS.iter().chain(controls.iter()) {
        let case = qualification["cases"]
            .as_array()
            .unwrap()
            .iter()
            .find(|case| case["case_id"] == *id)
            .unwrap();
        assert_eq!(case["disposition"], "admitted-for-execution");
        let fingerprints = case["typescript_run_fingerprints"].as_array().unwrap();
        assert_eq!(fingerprints.len(), 2);
        assert!(fingerprints
            .iter()
            .all(|fingerprint| fingerprint
                == &case["typescript_observation"]["run_fingerprint_sha256"]));
        let expected = expected(case);
        let floor = EmitOptionFloor::SourceMapWithOptions;
        let mut runs = Vec::new();
        for attempt in 0..2 {
            let actual = observe(prepare(case, &corpus, floor));
            capture(id, floor, attempt, &actual, &expected);
            runs.push(actual);
        }
        assert_eq!(runs[0], runs[1], "{id}: deterministic preparation and emit");
        assert_eq!(
            runs[0]["observation"], expected,
            "{id}: complete command tuple"
        );
        let activity = &runs[0]["map_runtime_activity"];
        assert_eq!(activity["H2.6c"], 0, "{id}: no later map owner");
        assert_eq!(
            activity["H2.6b"].as_u64().unwrap() > 0,
            ROWS.contains(id),
            "{id}: embedded sources/roots use the existing H2.6b owner"
        );
        eprintln!("{id}: H2.6a projection EXACT x2");
    }
}

#[test]
fn map_option_directives_and_virtual_configs_match_typescript() {
    let witnesses = artifact("crates/compiler/tests/fixtures/h2-6a-map-option-projection.json");
    assert_eq!(witnesses["typescript"], "6.0.3");
    assert_eq!(witnesses["cases"].as_array().unwrap().len(), 31);
    let qualification = artifact("ratchets/h2-6a-qualification.v1.json");
    let corpus = load_recorded_execution_plans(&workspace()).unwrap();
    let template = recorded(
        qualification["cases"]
            .as_array()
            .unwrap()
            .iter()
            .find(|case| case["case_id"] == ROWS[0])
            .unwrap(),
        &corpus,
    );
    for case in witnesses["cases"].as_array().unwrap() {
        assert_eq!(case["repetitions"], 2);
        for route in ["qualified", "recorded"] {
            let id = format!("{route}/{}", case["case_id"].as_str().unwrap());
            let floor = EmitOptionFloor::SourceMapWithOptions;
            let mut runs = Vec::new();
            for attempt in 0..2 {
                let prepared = if route == "qualified" {
                    qualified(&case["input"], floor)
                } else {
                    load_compiler_emit_with_option_floor(
                        &workspace(),
                        &witness_plan(&case["input"], template),
                        limits(),
                        floor,
                    )
                    .unwrap()
                };
                let actual = observe(prepared);
                capture(&id, floor, attempt, &actual, &case["observation"]);
                runs.push(actual);
            }
            assert_eq!(runs[0], runs[1], "{id}: deterministic preparation and emit");
            assert_eq!(
                runs[0]["effective_map_options"], case["effective_map_options"],
                "{id}: effective options"
            );
            assert_eq!(
                runs[0]["observation"], case["observation"],
                "{id}: complete command tuple"
            );
            eprintln!("{id}: H2.6a projection EXACT x2");
        }
    }
}

#[test]
#[ignore = "pre-change witness census; captures existing route behavior without promoting it"]
fn existing_witness_route_census() {
    let witnesses = artifact("crates/compiler/tests/fixtures/h2-6a-map-option-projection.json");
    let qualification = artifact("ratchets/h2-6a-qualification.v1.json");
    let corpus = load_recorded_execution_plans(&workspace()).unwrap();
    let template = recorded(
        qualification["cases"]
            .as_array()
            .unwrap()
            .iter()
            .find(|case| case["case_id"] == ROWS[0])
            .unwrap(),
        &corpus,
    );
    for case in witnesses["cases"].as_array().unwrap() {
        for (route, floor) in [
            ("qualified", EmitOptionFloor::MapFamily),
            ("recorded", EmitOptionFloor::MapFamily),
        ] {
            let id = format!("{route}/{}", case["case_id"].as_str().unwrap());
            let mut runs = Vec::new();
            for attempt in 0..2 {
                let prepared = if route == "qualified" {
                    qualified(&case["input"], floor)
                } else {
                    load_compiler_emit_with_option_floor(
                        &workspace(),
                        &witness_plan(&case["input"], template),
                        limits(),
                        floor,
                    )
                    .unwrap()
                };
                let actual = observe(prepared);
                capture(&id, floor, attempt, &actual, &case["observation"]);
                runs.push(actual);
            }
            assert_eq!(runs[0], runs[1]);
            eprintln!(
                "{id}: existing parity={}",
                runs[0]["observation"] == case["observation"]
            );
        }
    }
}
